[LTC] Upstream class Shape

alanwaketan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67672

Summary:
This commit Upstreams class Shape from lazy_tensor_staging branch.

Test Plan:
./build/bin/test_lazy --gtest_filter=ShapeTest.*

Differential Revision: [D32095478](https://our.internmc.facebook.com/intern/diff/D32095478)

Commit:
[LTC] Upstream class Shape

Summary:
This commit Upstreams class Shape from lazy_tensor_staging branch.

Test Plan:
WIP.

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32095478).

desertfire commented:
I thought yesterday's discussion was leaning towards getting rid of Shape? I am not opposing keeping Shape, and in fact I like it because that will make our merging easier. Just want to clarify so that we are all on the same page.

wconstab commented:
> I thought yesterday's discussion was leaning towards getting rid of Shape? I am not opposing keeping Shape, and in fact I like it because that will make our merging easier. Just want to clarify so that we are all on the same page.

No, we agreed to keep Shape yesterday, mainly for ease of having an operator== implementation.  But it'll still be a greatly simplified Shape (already removed most of the objectionable parts from it)

wconstab commented on torch/csrc/lazy/core/shape.h:
```diff
@@ -0,0 +1,78 @@
+    return c10::str(toString(at_element_type_), "[", c10::Join(",", dimensions_), "]");
+  }
+
+  c10::ScalarType at_element_type() const { return at_element_type_; }
```
should we rename this to `scalar_type()` to be consistent?

wconstab commented on torch/csrc/lazy/core/shape.h:
```diff
@@ -0,0 +1,78 @@
+  c10::ScalarType at_element_type() const { return at_element_type_; }
+  void set_element_type(at::ScalarType value);
+
+  int64_t rank() const { return dimensions_.size(); }
```
and change this to `dim()` similarly?

wconstab commented on torch/csrc/lazy/core/shape.h:
```diff
@@ -0,0 +1,78 @@
+#include <c10/util/Logging.h>
+
+namespace lazy_tensors {
+
```
not sure if you saw my message somewhere else, but can you also put the ShapeIndex class (From shape_util.h) here?  I moved it into shape.h in my big PR but that hasn't landed to staging yet.  ShapeIndex will be a dep of lowering_context unless we refactor that, but it's also pretty small and unoffensive looking

alanwaketan commented:
@wconstab This question just came out of my mind. Is Shape supposed to be backend specific?

alanwaketan commented:
> > @wconstab This question just came out of my mind. Is Shape supposed to be backend specific?
> 
> Trick question, I think.
> 
> lazy::Shape is not supposed to be backend specific. Meaning, it's designed to be used (a) by LTC code, in a backend-agnostic way, where it mirrors aten semantics faithfully; (b) as a way to pass aten shape information into backends
> 
> Backends may then take lazy::Shape and convert to e.g. xla::Shape (or subclass). At this time, they may add backend-specific information, but it shouldn't leak back out into the LTC APIs.

Then maybe we should have moved some of the shape related APIs from TsNode back to Node. That should be a separate task though.

The reason why I ask is more on whether the current directory is the right one to place the `Shape` implementation.

wconstab commented:
> > > @wconstab This question just came out of my mind. Is Shape supposed to be backend specific?
> > 
> > 
> > Trick question, I think.
> > lazy::Shape is not supposed to be backend specific. Meaning, it's designed to be used (a) by LTC code, in a backend-agnostic way, where it mirrors aten semantics faithfully; (b) as a way to pass aten shape information into backends
> > Backends may then take lazy::Shape and convert to e.g. xla::Shape (or subclass). At this time, they may add backend-specific information, but it shouldn't leak back out into the LTC APIs.
> 
> Then maybe we should have moved some of the shape related APIs from TsNode back to Node. That should be a separate task though.
> 
> The reason why I ask is more on whether the current directory is the right one to place the `Shape` implementation.

Yea, we might end up putting Shape back into Node; I'm not sure if we need to or not.  

But I think it's right to put shape.h in the main folder not the backend folder, if that's your question.

Commit:
Update on "[WIP][LTC] Upstream class Shape"


Summary:
This commit Upstreams class Shape from lazy_tensor_staging branch.

Test Plan:
WIP.

Differential Revision: [D32095478](https://our.internmc.facebook.com/intern/diff/D32095478)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32095478).

Commit:
Update on "[LTC] Upstream class Shape"


Summary:
This commit Upstreams class Shape from lazy_tensor_staging branch.

Test Plan:
./build/bin/test_lazy --gtest_filter=ShapeTest.*

Differential Revision: [D32095478](https://our.internmc.facebook.com/intern/diff/D32095478)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32095478).

alanwaketan commented:
Thanks @wconstab for approving this pull request. Waiting for the CI to land.
